src/sync/PushManager/index.js

```diff
@@ -0,0 +1,202 @@
+
+  // No return a PushManager if PUSH mode is not supported.
+  if (!checkPushRequirements(log)) return;
+
+  const pushEmitter = new EventEmitter();
+  const settings = context.get(context.constants.SETTINGS);
+  const storage = context.get(context.constants.STORAGE);
+  const sseClient = SSEClient.getInstance(settings);
+  const sseHandler = SSEHandlerFactory(pushEmitter);
+  sseClient.setEventHandler(sseHandler);
+
+  // map of hashes to user keys, to dispatch MY_SEGMENTS_UPDATE events to the corresponding MySegmentsUpdateWorker
+  const userKeyHashes = {};
+  // variable used on browser to reconnect only when a new client was added, saving some authentication and sse connections.
+  let connectForNewClient = false;
+
+  /** PushManager functions related to initialization */
+
+  const reauthBackoff = new Backoff(connectPush, settings.authRetryBackoffBase);
+  const sseReconnectBackoff = new Backoff(sseClient.reopen, settings.streamingReconnectBackoffBase);
+
+  let timeoutId = 0;
+
+  function scheduleTokenRefresh(issuedAt, expirationTime) {
+    // clear scheduled token refresh if exists (needed when resuming PUSH)
+    if (timeoutId) clearTimeout(timeoutId);
+
+    // Set token refresh 10 minutes before expirationTime
+    const delayInSeconds = expirationTime - issuedAt - SECONDS_BEFORE_EXPIRATION;
+
+    timeoutId = setTimeout(connectPush, delayInSeconds * 1000);
+  }
+
+  function connectPush() {
+    log.info('Connecting to push streaming.');
+
+    const userKeys = clientContexts ? Object.keys(clientContexts) : undefined;
+    authenticate(settings, userKeys).then(
+      function (authData) {
+        // restart backoff retry counter for auth and SSE connections, due to HTTP/network errors
+        reauthBackoff.reset();
+        sseReconnectBackoff.reset(); // reset backoff in case SSE conexion has opened after a HTTP or network error.
+
+        // emit PUSH_DISCONNECT if org is not whitelisted
+        if (!authData.pushEnabled) {
+          log.error('Streaming is not enabled for the organization. Switching to polling mode.');
+          pushEmitter.emit(PUSH_DISCONNECT); // there is no need to close sseClient (it is not open on this scenario)
+          return;
+        }
+
+        // don't open SSE connection if a new shared client was added, since it means that a new authentication is taking place
+        if (userKeys && userKeys.length < Object.keys(clientContexts).length) return;
+
+        // Connect to SSE and schedule refresh token
+        const decodedToken = authData.decodedToken;
+        sseClient.open(authData);
+        scheduleTokenRefresh(decodedToken.iat, decodedToken.exp);
+      }
+    ).catch(
+      function (error) {
+
+        sseClient.close(); // no harm if already disconnected
+        pushEmitter.emit(PUSH_DISCONNECT); // no harm if `PUSH_DISCONNECT` was already notified
+
+        if (error.statusCode) {
+          switch (error.statusCode) {
```

Comment:
I think we'll only have this error code with a special handling here correct? let's just use an if in that case.